### ISSUE TITLE
Circular padding does not work if the input tensor has no batch dimension

### ISSUE THREAD

**Issue by sebffischer** [mlverse/torch#940]
Something seems to be off with the circular padding. Even if `padding = 0` it throws an error. 

``` r
library(torch)
x1 = torch_randn(16, 1, 64, 64, requires_grad = FALSE)
x2 = torch_randn(1, 64, 64, requires_grad = FALSE)

conv1 = nn_conv2d(1, out_channels = 1, kernel_size = 3, padding_mode = "circular", padding = 0)
conv2 = nn_conv2d(1, out_channels = 1, kernel_size = 3, padding_mode = "zeros", padding = 0)

conv1(x1)$shape
#> [1] 16  1 62 62
conv1(x2)
#> Error in (function (self, pad, mode, value) : Invalid padding size, expected 2 but got 4
#> Exception raised from _pad_circular at ../aten/src/ATen/native/PadNd.cpp:92 (most recent call first):
#> frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6b (0x7f3cf8d239cb in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/lib/./libc10.so)
#> frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xce (0x7f3cf8d1f37e in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/lib/./libc10.so)
#> frame #2: at::native::_pad_circular(at::Tensor const&, c10::ArrayRef<long>) + 0x6a0 (0x7f3cd745e5f0 in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/lib/./libtorch_cpu.so)
#> frame #3: <unknown function> + 0x22544e6 (0x7f3cd80b44e6 in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/lib/./libtorch_cpu.so)
#> frame #4: at::_ops::_pad_circular::call(at::Tensor const&, c10::ArrayRef<long>) + 0x168 (0x7f3cd7aaf588 in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/lib/./libtorch_cpu.so)
#> frame #5: at::native::_pad_enum(at::Tensor const&, c10::ArrayRef<long>, long, c10::optional<double>) + 0x765 (0x7f3cd745d905 in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/lib/./libtorch_cpu.so)
#> frame #6: at::native::pad(at::Tensor const&, c10::ArrayRef<long>, c10::basic_string_view<char>, c10::optional<double>) + 0x126 (0x7f3cd745dd86 in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/lib/./libtorch_cpu.so)
#> frame #7: <unknown function> + 0x22545a8 (0x7f3cd80b45a8 in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/lib/./libtorch_cpu.so)
#> frame #8: at::_ops::pad::call(at::Tensor const&, c10::ArrayRef<long>, c10::basic_string_view<char>, c10::optional<double>) + 0x198 (0x7f3cd7ad3008 in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/lib/./libtorch_cpu.so)
#> frame #9: at::pad(at::Tensor const&, c10::ArrayRef<long>, c10::basic_string_view<char>, c10::optional<double>) + 0x5f (0x7f3ceed46f6f in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/lib/liblantern.so)
#> frame #10: _lantern_pad_tensor_intarrayref_cstringview_double + 0x16d (0x7f3ceec21bfc in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/lib/liblantern.so)
#> frame #11: cpp_torch_namespace_pad_self_Tensor_pad_IntArrayRef(XPtrTorchTensor, XPtrTorchIntArrayRef, XPtrTorchstring_view, XPtrTorchOptionaldouble) + 0x36 (0x7f3cf21f8c96 in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/libs/torchpkg.so)
#> frame #12: _torch_cpp_torch_namespace_pad_self_Tensor_pad_IntArrayRef + 0xd4 (0x7f3cf20ec4b4 in /home/sebi/R/x86_64-pc-linux-gnu-library/4.2/torch/libs/torchpkg.so)
#> frame #13: <unknown function> + 0xfe9c4 (0x7f3cfc7539c4 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #14: <unknown function> + 0xfef45 (0x7f3cfc753f45 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #15: <unknown function> + 0x13b399 (0x7f3cfc790399 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #16: Rf_eval + 0x88 (0x7f3cfc7ab2b8 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #17: <unknown function> + 0x15816f (0x7f3cfc7ad16f in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #18: Rf_applyClosure + 0x1a2 (0x7f3cfc7ae062 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #19: Rf_eval + 0x2af (0x7f3cfc7ab4df in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #20: <unknown function> + 0xc97bd (0x7f3cfc71e7bd in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #21: <unknown function> + 0x13b399 (0x7f3cfc790399 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #22: Rf_eval + 0x88 (0x7f3cfc7ab2b8 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #23: <unknown function> + 0x15816f (0x7f3cfc7ad16f in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #24: Rf_applyClosure + 0x1a2 (0x7f3cfc7ae062 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #25: <unknown function> + 0x143b6e (0x7f3cfc798b6e in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #26: Rf_eval + 0x88 (0x7f3cfc7ab2b8 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #27: <unknown function> + 0x15816f (0x7f3cfc7ad16f in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #28: Rf_applyClosure + 0x1a2 (0x7f3cfc7ae062 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #29: <unknown function> + 0x143b6e (0x7f3cfc798b6e in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #30: Rf_eval + 0x88 (0x7f3cfc7ab2b8 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #31: <unknown function> + 0x15816f (0x7f3cfc7ad16f in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #32: Rf_applyClosure + 0x1a2 (0x7f3cfc7ae062 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #33: <unknown function> + 0x143b6e (0x7f3cfc798b6e in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #34: Rf_eval + 0x88 (0x7f3cfc7ab2b8 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #35: <unknown function> + 0x15816f (0x7f3cfc7ad16f in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #36: Rf_applyClosure + 0x1a2 (0x7f3cfc7ae062 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #37: <unknown function> + 0x143b6e (0x7f3cfc798b6e in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #38: Rf_eval + 0x88 (0x7f3cfc7ab2b8 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #39: <unknown function> + 0x15816f (0x7f3cfc7ad16f in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #40: Rf_applyClosure + 0x1a2 (0x7f3cfc7ae062 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #41: Rf_eval + 0x2af (0x7f3cfc7ab4df in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #42: <unknown function> + 0x156d3c (0x7f3cfc7abd3c in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #43: <unknown function> + 0x1571e8 (0x7f3cfc7ac1e8 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #44: <unknown function> + 0x141aed (0x7f3cfc796aed in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #45: Rf_eval + 0x88 (0x7f3cfc7ab2b8 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #46: <unknown function> + 0x156d3c (0x7f3cfc7abd3c in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #47: Rf_eval + 0x5f8 (0x7f3cfc7ab828 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #48: <unknown function> + 0x1229a5 (0x7f3cfc7779a5 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #49: <unknown function> + 0x13b399 (0x7f3cfc790399 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #50: Rf_eval + 0x88 (0x7f3cfc7ab2b8 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #51: <unknown function> + 0x15816f (0x7f3cfc7ad16f in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #52: Rf_applyClosure + 0x1a2 (0x7f3cfc7ae062 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #53: <unknown function> + 0x143b6e (0x7f3cfc798b6e in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #54: Rf_eval + 0x88 (0x7f3cfc7ab2b8 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #55: <unknown function> + 0x15816f (0x7f3cfc7ad16f in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #56: Rf_applyClosure + 0x1a2 (0x7f3cfc7ae062 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #57: <unknown function> + 0x143b6e (0x7f3cfc798b6e in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #58: Rf_eval + 0x88 (0x7f3cfc7ab2b8 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #59: <unknown function> + 0x15816f (0x7f3cfc7ad16f in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #60: Rf_applyClosure + 0x1a2 (0x7f3cfc7ae062 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #61: Rf_eval + 0x2af (0x7f3cfc7ab4df in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #62: <unknown function> + 0x159de3 (0x7f3cfc7aede3 in /opt/R/4.2.2/lib/R/lib/libR.so)
#> frame #63: Rf_eval + 0x57b (0x7f3cfc7ab7ab in /opt/R/4.2.2/lib/R/lib/libR.so)
```

<sup>Created on 2022-12-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by skeydan** [comment]
I don't have insight into why circular padding has the restrictions it has, but from my experiments (see below) my guess is that the PyTorch (or maybe already Lua Torch) developers have built in an explicit assumption: that you want to use circular padding **only on the last dimension, if it's 1d (e.g., time) series**, and on the **final two, if it's 2d (an image)**.

Here a 1d series would always have to be encoded as a 3d tensor (what you put in to `nn_conv1d()`), and an image, as 4d ((what you put in to `nn_conv2d()`). **That is, the batch dimension always needs to be present** (_if_ you use circular padding).

That logic is encoded here: https://github.com/pytorch/pytorch/blob/fd3a7264ae80332ba5ec8f60446e6d7a2c2276c1/aten/src/ATen/native/PadNd.cpp#L109
It only applies to circular padding, evidently, which is why it does not appear in the documentation for Python `nn.functional.pad` (and thus not in R `nnf_pad`, either). (I also don't think we should add it on our side, since it could be an implementation detail on the Python side, seeing how it's not mentioned anywhere in the docs.)

Here's why I think what I wrote above:

```
# 2d, 3d, 4d input
x2 <- torch_randn(c(2, 2))
x3 <- torch_randn(c(2, 2, 2))
x4 <- torch_randn(c(2, 2, 2, 2))

### try 2d with padding = c(1, 1), which should pad the last dimension, on both sides
nnf_pad_circular(x2, padding = c(1, 1))

# Error in nnf_pad_circular(x2, padding = c(1, 1)) : 
# Invalid padding size, expected 0 but got 2

# looking at the C++ check gives
# own_size = 2
# ndim = own_size - 2 = 0
# padding_size = 2
# check: padding_size + 4 = own_size * 2
# fails because 2 + 4 != 2 * 2
# error msg
# "Invalid padding size, expected ", ndim * 2 = 0, " but got ", padding.size() = 2);

### conclusion: can never work for 2d tensor


### now try 3d with that same padding
nnf_pad_circular(x3, padding = c(1, 1))
# and works

### does 3d agree to pad both dimensions 2 and 3?
nnf_pad_circular(x3, padding = c(1, 1, 1, 1))
# no, since
# 4 + 4 != 3 * 2
# 2 4

### conclusion: "time series" can only be padded in the last dimension
#   for circular that is, not for constant!


### now pad 4d ("image")
nnf_pad_circular(x4, padding = c(1, 1))
# fails
# cannot pad one spatial dim only (if circular)
# but this works:
nnf_pad_circular(x4, padding = c(1, 1, 1, 1))
```

Let me know if this helps :-)

**Comment by sebffischer** [comment]
Yes that helped! :)

**Comment by skeydan** [comment]
great!
